File: docs/2026-II/AyG PUJ/C1/codigo/listas.cpp
// Clase 1 - Repaso de estructuras de datos y notacion asintotica
// Listas: std::deque mantiene referencias a los dos extremos,
// de modo que agregar y quitar por cualquiera de ellos cuesta O(1).
//
// Compilar y ejecutar:
//   g++ -std=c++17 -o listas listas.cpp && ./listas

#include <deque>
#include <iostream>
#include <string>

using namespace std;

void imprimir(const string &etiqueta, const deque<int> &l) {
    cout << etiqueta << " = [";
    for (size_t i = 0; i < l.size(); i++) {
        cout << l[i];
        if (i + 1 < l.size()) {
            cout << ", ";
        }
    }
    cout << "]" << endl;
}

void operaciones_en_los_extremos() {
    deque<int> l;

    l.push_back(10);    // O(1)
    l.push_front(5);    // O(1)
    l.push_back(20);
    imprimir("deque      ", l);

    cout << "front       = " << l.front() << endl;   // O(1)
    cout << "back        = " << l.back() << endl;    // O(1)
    cout << "size        = " << l.size() << endl;    // O(1)
    cout << "empty       = " << l.empty() << endl;   // O(1)

    l.pop_back();       // O(1)
    l.pop_front();      // O(1)
    imprimir("deque      ", l);
}

void operaciones_en_una_posicion() {
    // Llegar a la posicion k obliga a recorrer la secuencia: O(k).
    deque<int> l = {1, 2, 3, 4, 5};

    l.insert(l.begin() + 2, 99);   // O(k)
    imprimir("tras insert", l);

    l.erase(l.begin() + 2);        // O(k)
    imprimir("tras erase ", l);
}

int main() {
    cout << "Operaciones en los extremos" << endl;
    operaciones_en_los_extremos();

    cout << endl;
    cout << "Operaciones en una posicion intermedia" << endl;
    operaciones_en_una_posicion();

    return 0;
}

File: 2026-II/AyG PUJ/C1/codigo/listas.py
# Clase 1 - Repaso de estructuras de datos y notacion asintotica
# Listas: collections.deque mantiene referencias a los dos extremos,
# de modo que agregar y quitar por cualquiera de ellos cuesta O(1).

from collections import deque


def operaciones_en_los_extremos():
    l = deque()

    l.append(10)           # push_back,  O(1)
    l.appendleft(5)        # push_front, O(1)
    l.append(20)
    print('deque       =', list(l))

    print('front       =', l[0])      # O(1)
    print('back        =', l[-1])     # O(1)
    print('size        =', len(l))    # O(1)
    print('empty       =', len(l) == 0)

    ultimo = l.pop()       # pop_back,   O(1)
    primero = l.popleft()  # pop_front,  O(1)
    print('pop()       =', ultimo)
    print('popleft()   =', primero)
    print('deque       =', list(l))


def operaciones_en_una_posicion():
    # Llegar a la posicion k obliga a recorrer la secuencia: O(k).
    l = deque([1, 2, 3, 4, 5])

    l.insert(2, 99)        # insert(pos, x), O(k)
    print('tras insert =', list(l))

    del l[2]               # del l[pos],     O(k)
    print('tras del    =', list(l))


if __name__ == '__main__':
    print('Operaciones en los extremos')
    operaciones_en_los_extremos()

    print()
    print('Operaciones en una posicion intermedia')
    operaciones_en_una_posicion()
